### PR TITLE
New resource: `azurerm_databricks_access_connector`

### LINKED ISSUE #18518: Support for Databricks Access Connector
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Want to explore what is the work required to support Databricks Access Connector natively from `azurerm`

Currently, because the resource is not available from azurerm, we have to manage it through `azapi`.

```hcl
resource "azapi_resource" "ext_access_connector" {
  type      = "Microsoft.Databricks/accessConnectors@2022-04-01-preview"
  name      = "ext-databricks-mi"
  location  = data.azurerm_resource_group.this.location
  parent_id = data.azurerm_resource_group.this.id
  identity {
    type = "SystemAssigned"
  }
  body = jsonencode({
    properties = {}
  })
}
```

### New or Affected Resource(s)/Data Source(s)

azurerm_databricks_?

### Potential Terraform Configuration

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by favoretti**
Also upgrade Databricks SDK to `2022-04-01-preview` 
I added acceptance tests, but feature is in private preview.

Fixes #18518

### CODE REVIEW

**Comment by nfx** [comment]
Thank you so much, @favoretti !

**Comment by favoretti** [comment]
@tombuildsstuff One catch about this resource. Apparently SPN auth doesn't work with this API, only user-based auth. Wonder what to do about it..

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this @favoretti! Since this is still in private preview I expected the tests to fail when trying to create the connector. Unfortunately it's created the connector in a failed state and currently can't be deleted. I'm going to mark this PR as blocked for the time being while we investigate this further.

Also just as an aside all resources within the provider will eventually need to be converted to typed resources for the gradual shift to using Pandora generated ones. Our [contributor docs have a section on how to add new typed resources](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-new-resource.md) which will hopefully be helpful for future contributions 🙂.

**Comment by favoretti** [comment]
Thanks @stephybun. I completely blacked out on typed resources, I'll rewrite tomorrow, sorry about that. 

That failed connector will eventually be deletable. But the real reason it fails is that API doesn't work with SPN auth. I'm working with @nfx and team to see what the ETA is on supporting that. 

If persona authentication is used - resource works.

**Comment by nfx** [comment]
@favoretti You say it's the same style of issue as creating the AKV scope for Databricks? 🫣

**Comment by favoretti** [comment]
@stephybun rewrote this, SPN thing is still under investigation, but can you please have another look code-wise? Thanks!

**Comment by favoretti** [comment]
> @favoretti You say it's the same style of issue as creating the AKV scope for Databricks? 🫣

@nfx Don't think so. AKV scope talks to Databricks DataPlane API, i.e. directly to the instance. This call is directed at management.azure.com, i.e. ARM API. Although you folks would know better, but I think this one should be easier to address.

**Comment by favoretti** [comment]
I hear SPN auth should be fixed. I'll retest today.

**Comment by favoretti** [comment]
Local acceptance tests pass, running TC now.

```
$ TF_ACC=1 go test -v ./internal/services/databricks -timeout=1000m -run='TestAccDatabricksAccessConnector'
=== RUN   TestAccDatabricksAccessConnector_basic
=== PAUSE TestAccDatabricksAccessConnector_basic
=== RUN   TestAccDatabricksAccessConnector_requiresImport
=== PAUSE TestAccDatabricksAccessConnector_requiresImport
=== CONT  TestAccDatabricksAccessConnector_basic
=== CONT  TestAccDatabricksAccessConnector_requiresImport
--- PASS: TestAccDatabricksAccessConnector_basic (164.31s)
--- PASS: TestAccDatabricksAccessConnector_requiresImport (184.18s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/databricks	185.333s
```

**Comment by favoretti** [comment]
Failed test isn't related to either upgrade or the changes and seems to be transient. All the rest passes.
image

**Comment by favoretti** [comment]
@stephybun should be good to go for another review and potential merge. Would you take a look? Thanks a lot!

File: internal/services/databricks/databricks_access_connector_resource.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+			}
+
+			var state AccessConnectorResourceModel
+			if err := metadata.Decode(&state); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			existing, err := client.Get(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("reading Access Connector %s: %v", id, err)
```
Comment (stephybun) [inline internal/services/databricks/databricks_access_connector_resource.go:128]:
```suggestion
				return fmt.Errorf("reading %s: %v", id, err)
```

File: internal/services/databricks/databricks_access_connector_resource.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+			}
+
+			if metadata.ResourceData.HasChange("tags") {
+				existing.Model.Tags = &state.Tags
+			}
+
+			_, err = client.CreateOrUpdate(ctx, *id, *existing.Model)
+			if err != nil {
+				return fmt.Errorf("updating Access Connector %s: %+v", id, err)
```
Comment (stephybun) [inline internal/services/databricks/databricks_access_connector_resource.go:137]:
```suggestion
				return fmt.Errorf("updating %s: %+v", id, err)
```

File: internal/services/databricks/databricks_access_connector_resource.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+			}
+
+			client := metadata.Client.DataBricks.AccessConnectorClient
+
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				if !response.WasNotFound(resp.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+				return fmt.Errorf("while checking for Access Connectors's %q existence: %+v", id.ConnectorName, err)
```
Comment (stephybun) [inline internal/services/databricks/databricks_access_connector_resource.go:162]:
```suggestion
				return fmt.Errorf("retrieving %s: %+v", *id, err)
```

File: internal/services/databricks/databricks_access_connector_resource.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+			if err != nil {
+				return fmt.Errorf("while parsing resource ID: %+v", err)
+			}
+
+			client := metadata.Client.DataBricks.AccessConnectorClient
+
+			_, err = client.Delete(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("while removing Access Connector %q: %+v", id.ConnectorName, err)
```
Comment (stephybun) [inline internal/services/databricks/databricks_access_connector_resource.go:198]:
```suggestion
				return fmt.Errorf("deleting %s: %+v", *id, err)
```

**Review by stephybun — COMMENTED** [review]
A few minor comments on the error messages, other this is good to go!

**Comment by favoretti** [comment]
Thanks for the suggestions, @stephybun, committed, once unit tests complete, we should be gtg.

**Review by stephybun — APPROVED** [review]
Thanks @favoretti LGTM 🧋

**Comment by nfx** [comment]
🥳

**Comment by mvvsubbu** [comment]
> Thanks for this @favoretti! Since this is still in private preview I expected the tests to fail when trying to create the connector. Unfortunately it's created the connector in a failed state and currently can't be deleted. I'm going to mark this PR as blocked for the time being while we investigate this further.
> 
> Also just as an aside all resources within the provider will eventually need to be converted to typed resources for the gradual shift to using Pandora generated ones. Our [contributor docs have a section on how to add new typed resources](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-new-resource.md) which will hopefully be helpful for future contributions 🙂.


I am from Databricks Product group and I see the conversation about creation failing with SPN and Private Preview of the resource. Let me clarify those two issues here.

1.  SPN support for Access Connector is enabled in mid of October.
2. Access Connector Resource is in Public Preview (**Not Private Preview**).


We have received a customer escalation about the resource being documented as Private Preview. Can you help us understand the process to update this to Public Preview. 

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/databricks_access_connector

**Pull request merged**